drgn/cli.py

```diff
@@ -0,0 +1,151 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# SPDX-License-Identifier: LGPL-2.1-or-later
+"""Public interface to drgn's CLI"""
+import atexit
+import builtins
+import code
+import importlib
+import os
+import os.path
+import readline
+import shutil
+import sys
+from typing import Any, Dict, Optional
+
+import drgn
+from drgn.internal.rlcompleter import Completer
+
+__all__ = (
+    "drgn_version_header",
+    "run_cli",
+)
+
+
+def drgn_version_header() -> str:
+    """
+    Return the version header printed at the beginning of a Drgn session.
+    """
+    python_version = ".".join(str(v) for v in sys.version_info[:3])
+    libkdumpfile = f'with{"" if drgn._with_libkdumpfile else "out"} libkdumpfile'
+    return f"drgn {drgn.__version__} (using Python {python_version}, elfutils {drgn._elfutils_version}, {libkdumpfile})"
+
+
+def displayhook(value: Any) -> None:
+    if value is None:
+        return
+    setattr(builtins, "_", None)
+    if isinstance(value, drgn.Object):
+        text = value.format_(columns=shutil.get_terminal_size((0, 0)).columns)
+    elif isinstance(value, (drgn.StackFrame, drgn.StackTrace, drgn.Type)):
+        text = str(value)
+    else:
+        text = repr(value)
+    try:
+        sys.stdout.write(text)
+    except UnicodeEncodeError:
+        encoded = text.encode(sys.stdout.encoding, "backslashreplace")
+        if hasattr(sys.stdout, "buffer"):
+            sys.stdout.buffer.write(encoded)
+        else:
+            text = encoded.decode(sys.stdout.encoding, "strict")
+            sys.stdout.write(text)
+    sys.stdout.write("\n")
+    setattr(builtins, "_", value)
+
+
+def run_cli(
+    prog: drgn.Program,
+    additional_globals: Optional[Dict[str, Any]] = None,
```

Comment:
I actually initially had a `globals` and `additional_globals` which mirrored the banner args, but I decided against it because it was too cluttered.

I personally don't see the use case for too much flexibility here, because all you gain (vs `additional_globals`) is the ability to delete from the globals dict, or clear it altogether. Plus, it's a bit extra code to write for the common case of adding a few globals.

But I guess, flexibility should be preferred over brevity here. I wanted this API so I could do stuff beyond what the original design allowed, so I probably shouldn't assume to know how this will be used. So I'm on board with `globals_func` or some name along those lines.